services/gateway/src/lib.rs

```diff
@@ -124,14 +124,75 @@ enum RequestKind {
     RecoverAccount,
 }
 
+#[derive(Debug, Clone, thiserror::Error, ToSchema)]
+pub(crate) enum GatewayError {
+    #[error("Authenticator already exists")]
+    AuthenticatorAlreadyExists,
+    #[error("Transaction reverted on-chain (tx: {0})")]
+    TransactionReverted(String),
+    #[error("Transaction confirmation error: {0}")]
+    ConfirmationError(String),
+    #[error("Batcher unavailable")]
+    BatcherUnavailable,
+    #[error("Pre-flight check failed: {0}")]
+    PreFlightFailed(String),
+    #[error("{0}")]
+    Unknown(String),
+}
+
+impl Serialize for GatewayError {
+    fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+    where
+        S: serde::Serializer,
+    {
+        let code = match self {
+            GatewayError::AuthenticatorAlreadyExists => "AUTHENTICATOR_ALREADY_EXISTS",
+            GatewayError::TransactionReverted(_) => "TRANSACTION_REVERTED",
+            GatewayError::ConfirmationError(_) => "CONFIRMATION_ERROR",
+            GatewayError::BatcherUnavailable => "BATCHER_UNAVAILABLE",
+            GatewayError::PreFlightFailed(_) => "PRE_FLIGHT_FAILED",
+            GatewayError::Unknown(_) => "UNKNOWN",
+        };
+        serializer.serialize_str(code)
+    }
+}
+
+impl GatewayError {
+    pub(crate) fn from_contract_error(error: &str) -> Self {
+        // AuthenticatorAddressAlreadyInUse(address) selector: 0x218170d3
+        if error.contains("0x218170d3") {
+            return GatewayError::AuthenticatorAlreadyExists;
+        }
```

Comment:
there's a few additional errors we'd want to parse:
1. AuthenticatorDoesNotExist
2. MismatchedSignatureNonce
3. PubkeyIdInUse
4. PubkeyIdOutOfBounds
5. AuthenticatorDoesNotBelongToAccount